Firestore/Source/Local/FSTLRUGarbageCollector.mm

```diff
@@ -0,0 +1,100 @@
+#import "Firestore/Source/Local/FSTLRUGarbageCollector.h"
+
+#import <queue>
+
+#import "Firestore/Source/Local/FSTMutationQueue.h"
+#import "Firestore/Source/Local/FSTQueryCache.h"
+
+const FSTListenSequenceNumber kFSTListenSequenceNumberInvalid = -1;
+
+class RollingSequenceNumberBuffer {
+ public:
+  RollingSequenceNumberBuffer(NSUInteger max_elements)
+      : max_elements_(max_elements), queue_(std::priority_queue<FSTListenSequenceNumber>()) {
+  }
+
+  RollingSequenceNumberBuffer(const RollingSequenceNumberBuffer &other) = delete;
+  RollingSequenceNumberBuffer(RollingSequenceNumberBuffer &other) = delete;
+
+  RollingSequenceNumberBuffer &operator=(const RollingSequenceNumberBuffer &other) = delete;
+  RollingSequenceNumberBuffer &operator=(RollingSequenceNumberBuffer &other) = delete;
+
+  void AddElement(FSTListenSequenceNumber sequence_number) {
+    if (queue_.size() < max_elements_) {
+      queue_.push(sequence_number);
+    } else {
+      FSTListenSequenceNumber highestValue = queue_.top();
+      if (sequence_number < highestValue) {
+        queue_.pop();
+        queue_.push(sequence_number);
+      }
+    }
+  }
+
+  FSTListenSequenceNumber max_value() const {
+    return queue_.top();
+  }
+
+  std::size_t size() const {
+    return queue_.size();
+  }
+
+ private:
+  std::priority_queue<FSTListenSequenceNumber> queue_;
+  const NSUInteger max_elements_;
+};
+
+@interface FSTLRUGarbageCollector ()
+
+@property(nonatomic, strong, readonly) id<FSTQueryCache> queryCache;
+
+@end
+
+@implementation FSTLRUGarbageCollector {
+}
+
+- (instancetype)initWithQueryCache:(id<FSTQueryCache>)queryCache {
+  self = [super init];
+  if (self) {
+    _queryCache = queryCache;
+  }
+  return self;
+}
+
+- (NSUInteger)queryCountForPercentile:(NSUInteger)percentile {
```

Comment:
It seems like every caller is going to need to

```
NSUInteger queryCount = [collector queryCountForPercentile:10];
FSTListenSequenceNumber sequenceNumber = [collector sequenceNumberForQueryCount:queryCount];
[collector removeQueriesUpThroughSequenceNumber:sequenceNumber:sequenceNumber
    liveQueries:liveQueries
    group:group];
```
Might as well just wrap that in a method and make that public interface to this?